lib/bulk/common.js

```diff
@@ -427,26 +460,560 @@ function handleMongoWriteConcernError(batch, bulkResult, ordered, err, callback)
+    const maxWriteBatchSize =
+      topology.isMasterDoc && topology.isMasterDoc.maxWriteBatchSize
+        ? topology.isMasterDoc.maxWriteBatchSize
+        : 1000;
+
+    // Get the write concern
+    var writeConcern = applyWriteConcern(
+      shallowClone(options),
+      { collection: collection },
+      options
+    );
+    writeConcern = writeConcern.writeConcern;
+
+    // Get the promiseLibrary
+    const promiseLibrary = options.promiseLibrary || Promise;
+
+    // Final results
+    const bulkResult = {
+      ok: 1,
+      writeErrors: [],
+      writeConcernErrors: [],
+      insertedIds: [],
+      nInserted: 0,
+      nUpserted: 0,
+      nMatched: 0,
+      nModified: 0,
+      nRemoved: 0,
+      upserted: []
+    };
+
+    // Internal state
+    bulk.s = {
+      // Final result
+      bulkResult: bulkResult,
+      // Current batch state
+      currentBatch: null,
+      currentIndex: 0,
+      // ordered specific
+      currentBatchSize: 0,
+      currentBatchSizeBytes: 0,
+      // unordered specific
+      currentInsertBatch: null,
+      currentUpdateBatch: null,
+      currentRemoveBatch: null,
+      batches: [],
+      // Write concern
+      writeConcern: writeConcern,
+      // Max batch size options
+      maxBatchSizeBytes: maxBatchSizeBytes,
+      maxWriteBatchSize: maxWriteBatchSize,
+      // Namespace
+      namespace: namespace,
+      // BSON
+      bson: bson,
+      // Topology
+      topology: topology,
+      // Options
+      options: options,
+      // Current operation
+      currentOp: currentOp,
+      // Executed
+      executed: executed,
+      // Collection
+      collection: collection,
+      // Promise Library
+      promiseLibrary: promiseLibrary,
+      // Fundamental error
+      err: null,
+      // check keys
+      checkKeys: typeof options.checkKeys === 'boolean' ? options.checkKeys : true
+    };
+
+    // bypass Validation
+    if (options.bypassDocumentValidation === true) {
+      bulk.s.bypassDocumentValidation = true;
+    }
+  }
+
+  /**
+   * Add a single insert document to the bulk operation
+   *
+   * @param {object} document the document to insert
+   * @throws {MongoError}
+   * @return {OrderedBulkOperation or UnorderedBulkOperation}
+   */
+  insert(document) {
+    if (this.s.collection.s.db.options.forceServerObjectId !== true && document._id == null)
+      document._id = new ObjectID();
+    return this.addToOperationsList(this, INSERT, document);
+  }
+
+  /**
+   * Initiate a find operation for an update/updateOne/remove/removeOne/replaceOne
+   *
+   * @method
+   * @param {object} selector The selector for the bulk operation.
+   * @throws {MongoError}
+   * @return {FindOperators}
+   */
+  bulkFind(selector, bulk) {
+    if (!selector) {
+      throw toError('Bulk find operation must specify a selector');
+    }
+
+    // Save a current selector
+    bulk.s.currentOp = {
+      selector: selector
+    };
+  }
+
+  /**
+   * @method
```

Comment:
While we're here, please document what isn't documented. A description of the operation, its parameters, and return values